OpenRA.Mods.Common/Traits/BotModules/McvExpansionManagerBotModule.cs

```diff
@@ -0,0 +1,822 @@
+						{
+							var sdistance = (indiceCenter - location).LengthSquared;
+							if (sdistance <= indiceSideLengthSquare)
+							{
+								if (isAlly)
+									attraction -= indiceSideLengthSquare;
+								else
+									attraction -= indiceSideLengthSquare << 1;
+							}
+						}
+
+						foreach (var (othermcv, dest) in activeMCVs)
+						{
+							if (dest == indiceCenter && othermcv != mcv)
+								attraction -= indiceSideLengthSquare << 1;
+						}
+
+						if (!allowfallback)
+						{
+							var sdistance = (indiceCenter - mcv.Location).LengthSquared;
+							if (sdistance <= indiceSideLengthSquare)
+								attraction -= indiceSideLengthSquare << 1;
+						}
+
+						if (attraction > cb_best)
+						{
+							cb_best = attraction;
+							cb_checkspot = indiceCenter;
+							cb_suitablespot = indiceCenter;
+						}
+					}
+
+					return (cb_suitablespot ?? mcv.Location, cb_best, cb_checkspot);
+
+				/*
+				 * CheckResource mode considers the distance to current MCV, ally construction yard & refinery within range,
+				 * Attaction has a base value of:
+				 * 1. if not Mobile: indiceSideLengthSquare >> 4 (1/16 of the maximum distance weight, = 0.25 of the maximum euclid distance in map)
+				 * 2. if Mobile: indiceSideLengthSquare >> 3 (1/8 of the maximum distance weight, ≈ 0.35 of the maximum euclid distance in map)
+				 */
+				case BotMcvExpansionMode.CheckResource:
+
+					var cr_refinarylocs = world.ActorsHavingTrait<Refinery>()
+						.Where(a => a.Owner == player && resourceMapModule.Info.RefineryTypes.Contains(a.Info.Name))
+						.Select(a => (a.Location, a.Owner != player))
+						.ToArray();
+
+					var cr_conyardlocs = world.ActorsHavingTrait<Building>()
+						.Where(a => a.Owner.IsAlliedWith(player) && Info.ConstructionYardTypes.Contains(a.Info.Name))
+						.Select(a => (a.Location, a.Owner != player))
+						.ToArray();
+
+					// We only take indice has more than half of average indice value (in weight calculation), to skip the indice with very poor resource
+					// when failedAttempts is acceptable.
+					var thresholdRes = 0;
+					for (var i = 0; i < resourceMapModule.GetIndicesLength(); i++)
+					{
+						var resourceCellCounts = resourceMapModule.GetIndice(i).ResourceCellsCount;
+						thresholdRes += (indiceSideLengthSquare >> 1) - Math.Abs(resourceCellCounts - (indiceSideLengthSquare >> 1));
+					}
+
+					thresholdRes = (thresholdRes / resourceMapModule.GetIndicesLength()) >> 1;
+
+					CPos? cr_suitablespot = null;
+					CPos? cr_checkspot = null;
+					var cr_best = int.MinValue;
+
+					for (var i = 0; i < resourceMapModule.GetIndicesLength(); i++)
+					{
+						var indiceCenter = resourceMapModule.GetIndice(i).IndiceCenter;
+						var resourceCellsCount = resourceMapModule.GetIndice(i).ResourceCellsCount;
+						var resourceCellsCenter = resourceMapModule.GetIndice(i).ResourceCellsCenter;
+						var resourceCreatorLocs = resourceMapModule.GetIndice(i).ResourceCreatorLocs;
```

Comment:
```suggestion
						var index = resourceMapModule.GetIndice(i);
						var indiceCenter = index.IndiceCenter;
						var resourceCellsCount = index.ResourceCellsCount;
						var resourceCellsCenter = index.ResourceCellsCenter;
						var resourceCreatorLocs = index.ResourceCreatorLocs;
```

Maybe let's reuse the result